### PR TITLE
functional_collectives: add first differentiable collective -- all_to_all_single_grad

### PR DESCRIPTION
**Pull request by d4l3k**
This adds the differentiable collective -- all_to_all_single_grad. This is the initial proof of concept PR and I will be adding the remaining collectives in follow up PRs.

This adds a new function called `all_to_all_single_autograd` which is the autograd variant of `all_to_all_single`. For backwards compatibility + initial testing we wanted to make the autograd variant separate to avoid regressions.

This uses `autograd::Function` to register an Autograd op that calls the original `_c10d_functional::all_to_all_single` via the dispatcher. This works with compile and inductor as opposed to the previous Python implementation that had issues. As this uses the existing `_c10d_functional` ops we don't need to register any meta functions or lowering.

To avoid cudaStream issues this explicitly calls `wait_tensor` in the backward method to ensure it runs under the same stream as the async operation. This hurts performance but can be alleviated potentially using `compile`.

Related work: https://github.com/pytorch/torchrec/blob/main/torchrec/distributed/comm_ops.py

Test plan:

```
pytest test/distributed/test_functional_api.py -k test_all_to_all_single_compile
pytest test/distributed/test_functional_api.py
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: test/distributed/test_functional_api.py [hunk 1]
```diff
@@ -781,6 +781,44 @@ def test_dead_wrapper_plus_view(self):
...
+        super().setUp()
+        self._spawn_threads()
+
+    @property
+    def world_size(self):
+        return 2
+
+    @parametrize("compile", [True, False])
+    def test_all_to_all_single(self, compile: bool = True) -> None:
+        group = "0"
```
Comment (yifuwang) [inline test/distributed/test_functional_api.py:796]: nit: group names are subject to change. It's better to get it from `dist.group.WORLD` or just pass `dist.group.WORLD` to the funcol API.

**Comment by yifuwang** [comment]
Looks great! Can you add a `FileCheck` in `test_c10d_functional_native.py` test similiar to https://github.com/pytorch/pytorch/blob/c3d37a88edd5fa37e1c1d1fbf42915c85ea499f4/test/distributed/test_c10d_functional_native.py#L724?

**Review by yifuwang — APPROVED** [review]
Thanks a lot for considering all the suggestions and making it work. This looks great!

Let's continue the discussion on how the differentiable version should be exposed to users (via a separate API vs. a single API that supports differentiation). Though I don't think it needs to be a blocker for this PR.

**Comment by Chillee** [comment]
In general, does it makes sense to make distributed comm ops differentiable? I always thought that it depended on what the sharding configuration was.

i.e.

fw: all-gather => bw: if replicated, then just split. If it's a partial sum, then, it's a reduce-scatter. If it's sharded, then it's a no-op?

File: torch/distributed/_functional_collectives.py [hunk 2]
```diff
@@ -625,8 +662,13 @@ def wrap(e: torch.Tensor):
         unwrapped_args = tree_map_only(AsyncCollectiveTensor, unwrap, args)
         unwrapped_kwargs = tree_map_only(AsyncCollectiveTensor, unwrap, kwargs)
 
-        # we don't wrap the result as it doesn't need to be waited on.
-        out = func(*unwrapped_args, **unwrapped_kwargs)
+        # If elem has a grad_fn we need to enable reentrant_dispatch so that
+        # grad is propagated correctly. Normally the grad_fn would be on
+        # AsyncCollectiveTensor but since we wrapped a tensor with a grad_fn we
+        # need this.
+        with torch.overrides.enable_reentrant_dispatch():
```
Comment (albanD) [inline torch/distributed/_functional_collectives.py:669]: cc @zou3519 how do you feel about this one?
In this case, the wrapper doesn't require grad or has a grad_fn, only the wrapped Tensor and so we run autograd on the underlying Tensor here.
Off the top of my head, I think this will go wrong:
- When mixed with other subclasses that are resolved earlier (like DTensor). Where they will run their autograd on the wrapper (thinking it doesn't require grad) and we will create two different autograd graph
- If anyone checks the requires_grad field of the wrapper when it is returned (like a custom autograd.Function), it will also return the wrong thing.

The proper fix here is to fix the constructor to use a differentiable constructor like is done for other wrapper subclasses. For example DTensor: https://github.com/pytorch/pytorch/blob/f61b04a1f0ed55aa3f9b75e1266e7a5dc71fc90d/torch/distributed/_tensor/api.py#L37

Comment (zou3519) [reply torch/distributed/_functional_collectives.py:669]: Yeah we don't want `enable_reentrant_dispatch` here

**Review by zou3519 — CHANGES_REQUESTED** [review]
See Alban's comment

File: torch/distributed/_functional_collectives.py (hunk 2 above)
Comment (d4l3k) [reply torch/distributed/_functional_collectives.py:669]: Thanks for pointing out those wrappers -- I was looking for a way to do this but couldn't find anything

**Comment by d4l3k** [comment]
@Chillee I think there's a lot of use cases where having differentiable comm ops are useful such as in the ring attention case. There's definitely a variety of different patterns/usecases but I think there's only a handful of such patterns per collective and we can have a couple of variants if just one isn't sufficient.

For implementing RingAttention having something like this makes it much easier for end users to write/iterate on -- advanced users can either disable grad or use the non-grad variants of them for now.

This is following some prior art in torchrec where they also have a use case for these https://github.com/pytorch/torchrec/blob/main/torchrec/distributed/comm_ops.py

**Comment by wconstab** [comment]
> In general, does it makes sense to make distributed comm ops differentiable? I always thought that it depended on what the sharding configuration was.

I think that when folks talk about adding autograd to collectives, they are assuming the inputs to those collectives during forward were replicated.

this is potentially dangerous as there is no way to check for this in general.  (And if the user is using a framework like DTensor to do the sharding, then DTensor itself is already managing the backwards and would just avoid using the differentiability of comm ops).

There are clear use cases for this, but, i am a bit worried about how to best segment the apis so that the automatic backwards doesn't kick in and bite people who weren't expecting it or don't understand it.

**Comment by d4l3k** [comment]
> Looks great! Can you add a `FileCheck` in `test_c10d_functional_native.py` test similiar to
> 
> https://github.com/pytorch/pytorch/blob/c3d37a88edd5fa37e1c1d1fbf42915c85ea499f4/test/distributed/test_c10d_functional_native.py#L724
> ?

done

**Comment by d4l3k** [comment]
> assuming the inputs to those collectives during forward were replicated

Even if the inputs are replicated there's no correctness issue AFAIK about backpropagating through them -- you'll waste some computation but shouldn't cause incorrect results

> best segment the apis so that the automatic backwards doesn't kick in and bite people who weren't expecting it or don't understand it

This PR adds a new op `all_to_all_single_grad` with the `_grad` suffix to better indicate the behavior. @yifuwang suggested this to make it more obvious and avoid any breaking changes. Though it's still up in the air whether that's the best long term API.

**Comment by wconstab** [comment]
> Even if the inputs are replicated there's no correctness issue AFAIK

my point was the opposite- i think you are defining the backward formulas with the assumption that the input on each rank to the collective is logically a 'replica'.  If instead the user manually had done sharding and then called a collective on a shard, you might want something else during backward to horace's point.

**Comment by yifuwang** [comment]
> In general, does it makes sense to make distributed comm ops differentiable? I always thought that it depended on what the sharding configuration was.

Hmm is it? While one may want to use different backward in different scenarios, from send and recv's perspective, I think the gradient flow path is unique? i.e. sending a param on fwd -> receiving the grad of the param on bwd. Broadcasting a param on fwd -> receiving the grads and accumulate them on bwd. Curious if I'm missing something.

**Comment by d4l3k** [comment]
I wrote an example integration on how to compose this with DTensor: https://github.com/pytorch/pytorch/pull/123685

**Comment by d4l3k** [comment]
Got offline agreement that this is the right direction -- landing
